### DISCUSSION TITLE
Modular product types

### DISCUSSION THREAD

**Discussion by joshuatf** [woocommerce/woocommerce#45724] [General]
The usage of templates in the new product editor and attempt at combining features from multiple product types into a single template have exposed some long-standing challenges in WooCommerce product types.

I previously explored the option of [deprecating product types](https://github.com/woocommerce/woocommerce/discussions/44814) in favor of the concept of "traits."  I've since opened a few more concepts to explore this option to see what problems we might face in this process.  My latest concept moves towards the ideas of "modules" in #45722 as pieces of functionaity that enhance the base product.

**Relevant POC PRs:**

https://github.com/woocommerce/woocommerce/pull/44657
https://github.com/woocommerce/woocommerce/pull/44585
https://github.com/woocommerce/woocommerce/pull/45635
https://github.com/woocommerce/woocommerce/pull/45722

## Why modules

There a few major benefits to moving away from product types:

1. We allow a combination of multiple product features from different types within a single product.  This can also be retroactively updated to declare compatibility or incompatibility.
2. Product templates (both in form and potentially frontend) become easier to manage as the compatibility engine handles turning on or off features within the template.  The template is purely for restructuring the product blocks, if necessary.
3. We get a more consistent interface returned from the product factory.  Instead of a class that may contain certain methods and not others, we always have the same set of methods, while product modules can be retrieved and cast as necessary.

## The strategy

In the latest PR, https://github.com/woocommerce/woocommerce/pull/45722, I explore the idea of creating a new product type `WC_Product_Modular` that acts as the new base product extensible with modules.  By creating a new product type, we:

1. Make this an opt-in feature and only expose it in the new editor.
2. Allow the existing product type classes to work as is without breaking existing code.
3. Let users convert a product back to one of the existing core product types if a compatibilty issue with an extension is found.

### Implementation

This implementation keeps modules isolated and more closely follows SOLID principles.  While earlier concepts explored integrating traits directly into the product class, we ultimately end up with an inconsistent interface depending on which modules are available.  With this concept modules get registered on the `woocommerce_product_modules` hook:

```php
$product_modules = apply_filters( 'woocommerce_product_modules', $product_modules );

foreach ( $product_modules as $module ) {
	if ( is_string( $module ) && class_exists( $module ) ) {
		$module::init();
		$this->product_modules[ $module::get_slug() ] = $module;
	}
}
```

Module instances get created on each product:

```php
class WC_Product_Modular extends WC_Product {
	public function __construct( $product = 0 ) {
		parent::__construct( $product );
		$modules          = WC()->product_modules()->get_all_modules();
		$this->data_store = WC_Data_Store::load( 'product-modular' );

		foreach ( $modules as $module_class_name ) {
            		$module                        = new $module_class_name( $this );
            		$module_slug                   = $module::get_slug();
            		$this->modules[ $module_slug ] = $module;
		}
	}
}
```

Modules then can be activated, accessed, or deactivated:

```php
$product = new WC_Product_Modular();
$product->activate_module( 'variable' );
$product->activate_module( 'subscription' );
$product->get_module( 'subscriptions' )->get_trial_length();
$product->get_module( 'variable' )->get_available_variations();
$product->deactivate_module( 'subscription' );
```

## Next steps

* Get input from @woocommerce/proton and @woocommerce/mothra to get some buy-in to this idea.
* Flush out any blockers, issues with backwards compatiblity, or non-starters to this idea

**Comment by pmcpinto** [comment]
Will this product type include all the Core features by default?
If an extension uses this product module, will a new type be introduced in the product type selector?

**Reply by joshuatf** [reply]
>Will this product type include all the Core features by default?

Yes, this would include all core features.

>If an extension uses this product module, will a new type be introduced in the product type selector?

The main purpose of this concept is to remove "type" selection for users and reduce the complexity of the UI.  In some cases, extensions may want to just introduce a toggle or inherently turn on/off a module based on interaction with a field (e.g., variations automatically turns on/off depending on whether variations exist).

However, the option always exists to include a new "template" in the dropdown, which can be what we think of as "types" or can be a cobination of preset data to help merchants more quickly create a product.

**Comment by barryhughes** [comment]
Small thing: might a classname be preferable to a slug? Depending on how `get_module` is implemented, this could be used to guarantee the return type (that is, if we ask for `foo::class` we can expect to get back either an instance of `foo` or else `false`).

```diff
- $product->get_module( 'subscriptions' );
+ $product->get_module( Subscriptions_Module::class );
```

**Reply by joshuatf** [reply]
I think this has its advantages.  The one downside to this is the need to check classes (at least on third party integrated modules) before retrieving a module, but this is a check that needs to occur at some point regardless.

**Comment by barryhughes** [comment]
If a module is not activated, does that prohibit interaction? 

For example, if a hypothetical downloads module is initially activated, and downloadable assets are added/registered, and then for some reason the module is *deactivated* ... can we still access the module and its methods, or would that be prohibited?

**Reply by joshuatf** [reply]
Good question.  In earlier concepts I explored both the idea that (some) methods could be restricted depending on whether or not the module was activated and also that all methods could passthrough from the product to the traits.

In the latest POC, I've strayed away from that idea because I think it leads to an inconsistent interface.  So in short, those methods should always be available and not dependent on module activation.

**Comment by joshuatf** [comment]
@daigo75 I see you left a thumbs down on this discussion and I understand you might have some concerns over backwards compatibility.

The current concept attempts to remove any issues with backwards compatibility by creating a separate type (making this new modular type opt-in and not affecting other types) but still allowing classic `get_type` checks for plugins that may be relying on that check.  This has worked in a (admittedly narrow) set of use cases with extensions, but I'd love to hear your feedback on where you think this might cause issues.

I want to stress that this is purely exploratory at the moment and not something we would implement if it would cause issues in backwards compatibility.  If you have use cases you're concerned about, it would be great to share those so we can discuss and make those cases are covered.

**Reply by daigo75** [reply]
I'm quite busy these days, and I regret that I don't have much time for a long conversation on the subject, so I will try to be brief. One use case I can see causing trouble is the reliance of the existing product classes. I've personally seen quite a few of plugins and customisations that rely on specific product classes, with their own properties and interfaces, and extend them. 

Something as simple as this:
```
class My_Product extends WC_Product_Simple {
  // Some stuff
}
```
Could stop working, and it would have to be rewritten to be compatible with the "old" structure and the new one. It could be a serious breaking change, like the one that they made in PHP 8.2 when they removed dynamic properties. In one PHP version, you can use them. In the next one (a **minor** update, a x.2) you can't. I voted against that change too, and twice as much because the PHP developer was on his way out. In that case, all the opposition was met with "_but you shouldn't have used dynamic properties_". The way I see it, it was perfectly fine to do use dynamic properties, at least until a single individual decided it wasn't, therefore "_you shouldn't do that_" wasn't a valid answer for me. They provided a "workaround" to enable dynamic properties on specific classes, **but they didn't take into account that one might not be able to modify the code of these classes**. Due to that, the workaround is practically useless, unless one has full control over the whole codebase (which rarely happens, if one works with 3rd party libraries or code, such as WooCommerce).

In short, I would like to avoid the same mess. To guarantee backward compatibility with the code that relies on the current architecture, the existing classes will have be preserved, with their existing interfaces and behaviour. If that's not possible, then the knock on effect will be very serious, and I don't see the benefit in the new approach.

**Reply by joshuatf** [reply]
Thanks for your thoughts, Diego!!

>Something as simple as this: class My_Product extends WC_Product_Simple {

I think your concern about this type of break is valid, but this would still work given these classes and types still exist.  While my earlier discussions and POCs did favor deprecating product types, the concept in this post creates a new type that is more modular.  Plugins can continue using types and existing classes as is.  Programmatically, very little has to change for extensions that are creating their own types unless they want to opt into the modular product experience.

My larger backwards compatibility concerns are around extensions doing something funky with extending existing type functionality in a manner that's less than ideal:

```php
function add_something_to_product( $product ) {
  if ( $product->get_type() === 'simple' ) { ... } // This would work fine with modular products

  if ( is_a( $product, 'WC_Product_Simple' ) ) { ... } // This would not work since the class would be `WC_Product_Modular`
}
```

I have yet to see this exact case in the wild, but I don't doubt that it might exist.  And if it does exist, I'd like to see what impact that has as it may be something that does not pose major issues or it may be a dealbreaker.  It may also be acceptable given we're working with a new type altogether.

> To guarantee backward compatibility with the code that relies on the current architecture, the existing classes will have be preserved, with their existing interfaces and behaviour.
✅ ✅ ✅

 >and I don't see the benefit in the new approach.

Not sure if this was a continuation of your previous thought and contigent on breaking interface, but just noting that a bigger part of this change is less about developer benefits (though I do think it pushes towards a more consistent interface long-term) and more about how this relates to merchant interface.

Product types offload a good amount of responsibility to merchants to figure out how to correctly configure products.  What has become abundantly clear in user testing is that the current UI is less than ideal and a major pain point in creating a better UI is type selection.

And again, none of that is to say that this is the best solution; it's just one idea that moves towards solving that issue.  The more cases we can discuss around it, the better we can vet out whether or not it's worth pursuing. 🙂

**Reply by daigo75** [reply]
I can tell you that `is_a()` and `instanceof` are in use. I used the latter them myself, back when there was no `get_type()` function, and the code is running smoothly to this day. 

In addition to that, functions could expect a `WC_Product_Simple` argument, or a `WC_Product_Variable`, or something like that.

If you plan to have a `WC_Product_Simple`, and other classes, that extend `WC_Product_Modular`, that might work. If not, that's going to be an issue.

Furthermore, building product types that may or may not have specific features will make it tricky to support them. My day to day work revolves around handling custom prices for products , and that can only be done by knowing what a product is and what it does. Simple products are handled in one way, variable in a slightly different way, and so on. Small digression: bundles, subscriptions, composites, etc should be handled by the respective authors, who know how these products work, even if that hasn't yet happened with some plugins. End of digression.

That would be about it, I think.

**Reply by joshuatf** [reply]
>If you plan to have a WC_Product_Simple, and other classes, that extend WC_Product_Modular, that might work. If not, that's going to be an issue.

There will still be a `WC_Product_Simple` class.  What other classes are you expecting to extend `WC_Product_Modular` or did I misunderstand you here?

>Furthermore, building product types that may or may not have specific features will make it tricky to support them.

We already have this to a large extent with the product factory which is why extensions check the type and/or class of a product to determine what features are available.  I don't believe that's inherently different than checking if a module is active on a class; it's just a different way of checking.  You just as easily be able to check by running `$product->is_module_active( Variable::class )`.

While there are many cases of product types that probably should not have been made types and can't easily retroactively convert to non-types, there are even more cases of plugins extended existing types with features that are not compatible with one another. 
 The advantage to the modular approach is being able to declare 2 "modules" or "traits" compatible or incompatible, solving a lot of these ongoing compatibility issues between third party plugins.

But I hear you on the legacy code and nuisance that adding another type check would represent.  I'm not yet convinced that the modular type is the right approach, but it is very clear that we need a better way of handling compatibility between product features (not just types).

**Comment by EdithAllison** [comment]
@joshuatf The use case that worries you is used here: https://brickslabs.com/create-a-customizable-ajax-add-to-cart-button-in-bricks-builder/
<img width="771" alt="Bildschirmfoto 2024-03-28 um 18 57 13" src="https://github.com/woocommerce/woocommerce/assets/19706428/c4e93c1b-65bf-41d0-9c7f-cdb059b66fee">

**Reply by joshuatf** [reply]
Thanks for the example, Edith!

In this particular example (which I think highlights the majority of cases), it seems like the functionality added when `itchy_is_simple_product()` is `true` would only apply to simple products.  So this would continue to work with simple products, but `itchy_is_simple_product ` would have to be extended if products of the new modular type want to be included.

I think this is good and expected behavior in most ways as it prevents breaking changes, but I can also see the case where it's a nuisance for developers to need to include one more product type check.  This is something we'll need to weigh out and see if the perceived UI benefits are worth this added cost.

**Comment by leewillis77** [comment]
I haven't reviewed the PRs in detail, but from a skim read of the proposal and discussions I'd say that I'm against this.

My principal objection would be that this will further fragment the platform, and doesn't provide any meaningful compatibility / smoothed adoption process. Some specific comments below:

> Make this an opt-in feature and only expose it in the new editor.

As I've commented elsewhere, the "new editor" is nowhere near suitable for use by third party extensions, so adding new features and (presumably) encouraging users to use that editor to get access to the new features fragments, and harms the platform.

> Allow the existing product type classes to work as is without breaking existing code.
> Let users convert a product back to one of the existing core product types if a compatibilty issue with an extension is found

While that may be considered "backwards compatible" in a purely technical sense, I don't believe that it provides any meaningful compatibility benefits. It will look "broken" to users and the finger will likely get pointed at 3rd party extensions. 

> The current concept attempts to remove any issues with backwards compatibility [...] still allowing classic get_type checks for plugins that may be relying on that check.

That isn't really providing any useful compatibility if `get_type()` returns something other than 'simple' for a simple-only 'modular' product.

> I have yet to see this exact case in the wild, but I don't doubt that it might exist. 

We definitely have extensions that switch their behaviour and implementation on either the return value of `get_type()` and/or the class type. Ideally, if a customer creates a 'simple' product in the new editor it would look and behave like WC_Product_Simple to code that doesn't specifically support modular products (both in terms of the specific class instance generated and the return from methods like `get_type()`). Ditto for 'variable' products looking/behaving like WC_Product_Variable. That would be backward compatible. Making all new simple or variable products look and behave different to current code isn't useful from a user/3rd party-dev point of view IMHO. 

> The main purpose of this concept is to remove "type" selection for users and reduce the complexity of the UI. 
> [...]
> noting that a bigger part of this change is less about developer benefits (though I do think it pushes towards a more consistent interface long-term) and more about how this relates to merchant interface.

Is this change (or something like it from a development point of view) really required just to remove a drop-down from the UI? It doesn't seem like it should be. 

> We already have this to a large extent with the product factory which is why extensions check the type and/or class of a product to determine what features are available. I don't believe that's inherently different than checking if a module is active on a class; it's just a different way of checking. 

Not sure how relevant it is (but it is a difference), but it would mean that you can't rely on PHP's type checking to enforce that you're receiving the type of product you expect to be receiving.

> The advantage to the modular approach is being able to declare 2 "modules" or "traits" compatible or incompatible, solving a lot of these ongoing compatibility issues between third party plugins.

I don't think I've ever come across a case of this - can you provide a specific example of a problem like this for context. In general I'd expect those to be solvable by the extensions themselves under the current system if they want to. If they don't know about the issue (or care), then they're unlikely to mark themselves incompatible either so not sure this really brings any real benefit.

**Reply by joshuatf** [reply]
Thanks for your comments, Lee!

>As I've commented elsewhere, the "new editor" is nowhere near suitable for use by third party extensions, so adding new features and (presumably) encouraging users to use that editor to get access to the new features fragments, and harms the platform.

I think this argument is mostly around the new product editor and not necessarily related to the modular product types.

I understand and agree with your point that the product editor should be fully ready to use by all extensions before any kind of change occurs.  I know it's missing a few key fields that you need for your extensions.

The discussion in this thread is a question on long-term strategy and I think it's best for the sake of discussion to separate concerns around the product editor into its respective thread. 

>While that may be considered "backwards compatible" in a purely technical sense, I don't believe that it provides any meaningful compatibility benefits. It will look "broken" to users and the finger will likely get pointed at 3rd party extensions.

I think this is a misunderstanding of the concept.  The modules in this concept can declare compatibility with one another.  If compatibility has not been declared, existing product type classes can continue to be used.

The user does not see any of this and types that have not declared compatibility continue to work as they previously did.

 >That isn't really providing any useful compatibility if get_type() returns something other than 'simple' for a simple-only 'modular' product.

I'm not quite understanding this statement.  If a product has a custom module enabled and `get_type()`  returns `custom-product-type`, are you saying this is not useful to plugins that may have been checking the value of `get_type()`?

>Is this change (or something like it from a development point of view) really required just to remove a drop-down from the UI? It doesn't seem like it should be.

If we're talking about removing product type selection (whether dropdown or some other form), then the answer is "maybe."  We do know with the current product type structure, we cannot remove product type selection and inherently set product types based on a product's other settings since product types would "compete" with one another.

This has been something extremely limiting for the design team so we're exploring long-term strategies that don't restrict the UI as much.  The approach in this thread is just an exploration and may not be the right one; other suggestions are always welcome.

>I don't think I've ever come across a case of this - can you provide a specific example of a problem like this for context.

Described in the product editor and in the example above (also see https://github.com/woocommerce/woocommerce/discussions/44814):

Core - Simple and Variable types
Extensions - Subscriptions and Variable types

The core example can be solved with quite a bit of duct-tape, but it means in the new product editor that extensions cannot reuse the product form templates provided.

WooCommerce Subscriptions also did solve the problem of needing variations by creating yet another product type "Variable Subscriptions" and duplicating a number of efforts made between those product types.

**Reply by leewillis77** [reply]
> > While that may be considered "backwards compatible" in a purely technical sense, I don't believe that it provides any meaningful compatibility benefits. It will look "broken" to users and the finger will likely get pointed at 3rd party extensions.
> 
> I think this is a misunderstanding of the concept. The modules in this concept can declare compatibility with one another. If compatibility has not been declared, existing product type classes can continue to be used.
> 
> The user does not see any of this and types that have not declared compatibility continue to work as they previously did.

It's definitely possible that there's a misunderstanding, but from what you've written the notion of 'backwards compatibility' is satisfied because users can 'convert' their product to one of the existing legacy types if an extension doesn't support the new approach. To the user that isn't backwards compatibility, that looks like something is broken and the assumption will be that it's the 3rd party extension's fault. 

That's especially relevant if even legacy-style products (e.g. 'simple' or 'variable' style products) created through the new system don't appear as the legacy types to existing code (see next point below)

> > That isn't really providing any useful compatibility if get_type() returns something other than 'simple' for a simple-only 'modular' product.
> 
> I'm not quite understanding this statement. If a product has a custom module enabled and `get_type()` returns `custom-product-type`, are you saying this is not useful to plugins that may have been checking the value of `get_type()`?

My question is, if with the new system in place, I create a product, setting just a title, description and price - ie the bare minimum, will the resulting product be WC_Product_Simple, and will a `get_type()` on that product return `simple`? If the answer is "no" then this change is not backwards-compatible. Ditto for a variable product. If I create a product with three variations, I'd expect it to create a WC_Product_Variable whose `get_type()` returns 'variable'. 

> > Is this change (or something like it from a development point of view) really required just to remove a drop-down from the UI? It doesn't seem like it should be.
> 
> If we're talking about removing product type selection (whether dropdown or some other form), then the answer is "maybe." We do know with the current product type structure, we cannot remove product type selection and inherently set product types based on a product's other settings since product types would "compete" with one another.

Isn't it just a process of elimination?
- If product has child products it's variable (and can't have grouped products or an external URL)
- If product has grouped products it's grouped (and can't have child products or an external URL)
- If product has an external URL it's external
- Otherwise, it's simple

> This has been something extremely limiting for the design team so we're exploring long-term strategies that don't restrict the UI as much. The approach in this thread is just an exploration and may not be the right one; other suggestions are always welcome.

I'd love to offer a suggestion, but honestly I'm still not sure I even understand the problem that you're trying to solve? As with many of the changes recently this seems to be re-engineering without any clear communication about the actual problem is with the current system. I've seen plenty of people protest that this isn't required, and not seen (apologies if I've missed it) anyone saying "Yes - this is what we need!". 

> > > The advantage to the modular approach is being able to declare 2 "modules" or "traits" compatible or incompatible, solving a lot of these ongoing compatibility issues between third party plugins.
> > 
> > I don't think I've ever come across a case of this - can you provide a specific example of a problem like this for context.
> 
> Described in the product editor and in the example above (also see https://github.com/woocommerce/woocommerce/discussions/44814):

I can't see any reference to any compatibility issues with third party plugins referenced - beyond that the Subscriptions plugin decided (for some reason, undocumented here) to create product types rather than layer functionality on top of the existing types and that is considered a "bad thing". 

> Core - Simple and Variable types
> Extensions - Subscriptions and Variable types
> 
> The core example can be solved with quite a bit of duct-tape

What example? What is the problem!? Simple and Variable products are already exclusive (by being separate types). What's the issue with that? Has anyone ever suggested that they want a product to be both Simple and Variable? That makes no sense...

> WooCommerce Subscriptions also did solve the problem of needing variations by creating yet another product type "Variable Subscriptions" and duplicating a number of efforts made between those product types.

So, surely the issue we should be solving is how an extension such as Subscriptions could have just added its functionality to the existing types and not had to create their own? Maybe I'm misunderstanding completely...

**Reply by joshuatf** [reply]
>It's definitely possible that there's a misunderstanding, but from what you've written the notion of 'backwards compatibility' is satisfied because users can 'convert' their product to one of the existing legacy types if an extension doesn't support the new approach. To the user that isn't backwards compatibility, that looks like something is broken and the assumption will be that it's the 3rd party extension's fault.

When we talk about converting to a legacy type, we're talking about allowing them to pick a type from a drop-down as stop-gap.  I'm not sure that a user would see this as any more broken than the existing interface as nothing has changed.  This just means that an extension hasn't opted into this new experience.

>ie the bare minimum, will the resulting product be WC_Product_Simple, and will a get_type() on that product return simple?

Yes to the `get_type()` return, but unfortunately I haven't' found a a reliable way return the class so any `get_class` calls on the product would not result in `WC_Product_Simple`.  This is good perspective to have and probably a reason that the approach proposed here will not work.

>Isn't it just a process of elimination?

You've identified one of the problem areas and I think this was also our hope when redesigning the editor interface, but unfortunately it's not this simple.

To back up for a moment, the problem we're trying to solve is removal of product type selection.  In user testing with the current experience and new designs, this was an area that was found to be challenging and unintuitive for new merchants.  Working under this assumption, the goal of the concept proposed in this post was to find a technical solution that allows removing of the type selection in favor of a more "implicit type" approach.

>If product has child products it's variable (and can't have grouped products or an external URL)
>If product has grouped products it's grouped (and can't have child products or an external URL)
>If product has an external URL it's external

We also initially believed that this approach was possible.  However, this does not work well in practice.  If all product type fields were suddenly added to a template, should adding content to one field automatically hide all other product type fields?  Ignoring the poor UI of that for a moment (which is certainly up for debate), there are many cases where fields across types can and should be compatible with one another (gift cards, variations, subscriptions).

This is commonly solved in the legacy experience by adding CSS classes throughout the code to show/hide fields as needed, but this requires that that one extension know far too much about another's fields.  The idea behind the modular approach is declaring compatibility with another module and have fields automatically shown/hide as needed without further intervention by extension developers.

That said, I think your point about being able to check classes at the very least invalidates this approach.  But long-term it would be nice to have a system that allows various "types" to be compatible one another and does not required extensions to manually show/hide other extension fields.

**Comment by jkohlbach** [comment]
I think Lee's comments are spot on for the most part. 

Having just written a new plugin that creates a new product type, I can say 100% that the current system is really not hard at all.

I made my new product type inherit the Simple and Variable product classes and then added whatever functionality I needed. 

There's a few small gotchas and things in this process that could be made simpler (like a better way to show/hide parts of the interface based on type) but for the most part, I was done in a few hours max.

I think the existing Product Types are an important concept in Woo and lots of people both use it in development and are "used to it" in terms of the way it works.

I also don't believe you'll be able to capture all the backwards compatibility for 3rd party extensions that are 1000% in on the exiting way product types work.

Talking about Subscriptions (which I believe is top of mind for core team) is a good example. They inherit the product classes, but they have so much additional functionality they're adding on simply by adding additional methods to a class. It's simple coding. If I'm reading this correctly, in order to get to modular, this would trigger a complete rewrite in how they handle their additional functionality.

I do believe the intentions here are pure, but from an outsider's perspective it seems the real issue the team is trying to avoid is not having to provide a way to give different experiences in the new product editor based on the type. I have brought up a few times on the new product editor channel that the exclusion of a type selector is a mistake.

**Reply by joshuatf** [reply]
Thanks for your thoughts, @jkohlbach!

>Having just written a new plugin that creates a new product type, I can say 100% that the current system is really not hard at all.
>I think the existing Product Types are an important concept in Woo and lots of people both use it in development and are "used to it" in terms of the way it works.

I understand the comfort in familiarity, though I want to stress that this is not necessarily how challenging the current experience is to develop, but how it limits design possibilities.

> If I'm reading this correctly, in order to get to modular, this would trigger a complete rewrite in how they handle their additional functionality.

It would not require a complete rewrite.  This would require some updates to how checks are made within extensions for various functionality to continue working.  E.g., `$product->get_type() === 'subscription'` -> `$product->has_module( 'subscription' )`

>I do believe the intentions here are pure, but from an outsider's perspective it seems the real issue the team is trying to avoid is not having to provide a way to give different experiences in the new product editor based on the type. I have brought up a few times on the new product editor channel that the exclusion of a type selector is a mistake.

This is about allowing more opportunities in design, that's correct.  Could you expand on why you think the exclusion of a type selector is a mistake?  That's certainly something that could be useful to discuss with our design teams.

**Comment by junixdev** [comment]
From a skim reading in the discussion here and on https://github.com/woocommerce/woocommerce/discussions/44814, I feel like this move is basically for just to make it easier for working on the following:
- subscription products
- the new product editor

I can also understand making the "virtual" and "downloadable" toggles as part of the modular aspect, but I don't really see how this would work in other use cases currently especially for other common product types that we know of like group, bookings, bundles, etc. as these product types are intended to be different from the simple and variable product types.

We haven't really dwelled yet with the new product editor, so can't really comment on the difficulty of implementing the current product code base for it. I'm all in for using traits as we also use it a lot in our plugins' code base, but adding this new product modular type that is intended to replace the current ones in the future (which surely will be long ways ahead), just adds unnecessary complexity and will be another thing for us to make our plugins compatible to.

Instead of `product types OR modules`, why not just implement this  as `product types AND modules` and just add the code to the current `WC_Product` abstract? this way we can keep the current product classes that we already have and maybe add the modules that we want to apply later on.

**Reply by jkohlbach** [reply]
+1 for adding to the current abstract, makes total sense and what I was trying to get at in [my comment above](https://github.com/woocommerce/woocommerce/discussions/45724#discussioncomment-10260392). The current system isn't broken.

**Reply by joshuatf** [reply]
Good points, @junixdev!

> I feel like this move is basically for just to make it easier for working on the following:
> * subscription products
> * the new product editor

You're not wrong about these.   There are many other product types that also would benefit from these changes, though I understand many extension developers don't have those needs.

>I can also understand making the "virtual" and "downloadable" toggles as part of the modular aspect, but I don't really see how this would work in other use cases currently especially for other common product types that we know of like group, bookings, bundles, etc. as these product types are intended to be different from the simple and variable product types.

For those cases, they simply don't declare compatibility with other modules and continue to work as they did as a "type."  A product type is essentially just a module/trait that is not compatible with other traits.  The goal of this system was to develop something that both allowed that functionality to continue to work while opening up the option for types to become compatible with one another (even retroactively).

>but adding this new product modular type that is intended to replace the current ones in the future (which surely will be long ways ahead), just adds unnecessary complexity and will be another thing for us to make our plugins compatible to.

Understood on the complexity.  We certainly don't want to needlesslly add more load to your plate and I know this is something that benefits some developers and is somewhat design dependent.

I think due to the backwards compatibility issues that @leewillis77 noted above on class checks, the proposed concept in this post is not something we will move forward with as this has potential to cause a lot of work for existing plugins.

**Comment by AashikP** [comment]
I've only skimmed through the discussions. Is my understanding right that the current product types that come with core are going to stay as they are, and the proposal is to add a new product type where custom functionality is added via traits? The `depreciating product type` sounds like we're no longer going to have any product type at all.

**Reply by joshuatf** [reply]
You're correct that with this proposal, the existing product types in core would remain untouched and any custom product types would continue to function.

The long-term goal would be to phase those product types out in favor of a single more modular type, however this needs to both gracefully degrade when a "legacy" product type is used and be backwards compatible with all extensions.  The former is solvable with this approach, while the latter still has some cases that are not completely backwards compatible.

For this reason, I'm not leaning heavily into this approach and we'll need to consider ways that don't upset the plugin compatibility.

**Reply by AashikP** [reply]
> The long-term goal would be to phase those product types out in favor of a single more modular type, however this needs to both gracefully degrade when a "legacy" product type is used and be backwards compatible with all extensions. The former is solvable with this approach, while the latter still has some cases that are not completely backwards compatible.

Thank you for clarifying!
